src/App.php

```diff
@@ -302,32 +307,87 @@ public function outputDebug($str)
         echo 'DEBUG:' . $str . '<br/>';
     }
 
+    /**
+     * Normalize HTTP headers to associative array with LC keys.
+     *
+     * @param string[] $headers
+     * @return string[]
+     */
+    protected function normalizeHeaders(array $headers): array
+    {
+        $res = [];
+        foreach ($headers as $k => $v) {
+            if (is_numeric($k) && ($p = strpos($v, ':')) !== false) {
+                $k = substr($v, 0, $p);
+                $v = substr($v, $p + 1);
+            }
+
+            $res[strtolower(trim($k))] = trim($v);
+        }
+
+        return $res;
+    }
+
+    /**
+     * @return $this
+     */
+    public function setResponseStatusCode(int $statusCode): self
+    {
+        $this->setResponseHeader(self::HEADER_STATUS_CODE, $statusCode);
+
+        return $this;
+    }
+
+    /**
+     * @return $this
+     */
+    public function setResponseHeader(string $name, string $value): self
+    {
+        $arr = $this->normalizeHeaders([$name => $value]);
+        $value = reset($arr);
+        $name = key($arr);
+
+        if ($value !== '') {
+            $this->response_headers[$name] = $value;
+        } else {
+            unset($this->response_headers[$name]);
+        }
+
+        return $this;
+    }
+
     /**
      * Will perform a preemptive output and terminate. Do not use this
      * directly, instead call it form Callback, jsCallback or similar
      * other classes.
      *
      * @param string|array $output Array type is supported only for JSON response
-     * @param string       $contentType
+     * @param string[]     $headers Content-type header must be always set or consider using App::terminateHTML() or App::terminateJSON() methods.
      *
      * @throws \atk4\core\Exception
      * @throws ExitApplicationException
      */
-    public function terminate($output = null, string $contentType = 'text/html'): void
+    public function terminate($output = '', array $headers = []): void
```

Comment:
@ibelar - will this break anything?

Reply:
It should not and null should never be allowed here. To not output anaything use nothing or empty string.